### PR TITLE
MAINT: fix usage of deprecated Numpy C API in optimize...

### PR DESCRIPTION
**Pull request by rgommers**
... and silence build warnings by defining `NPY_NO_DEPRECATED_API`.

### CODE REVIEW

**Comment by charris** [comment]
Was going to say LGTM, but there are problems with `_lbfgsbmodule.c`. There are other warnings also, have they always been there?

**Comment by rgommers** [comment]
@charris see https://github.com/numpy/numpy/pull/5309#issuecomment-68492561.

**Comment by rgommers** [comment]
Try again, now only for numpy >=1.10.

**Comment by rgommers** [comment]
Still an issue for 1.6.2: ``‘NPY_ARRAY_C_CONTIGUOUS’ undeclared``. It was only introduced in 1.7.0.  ``NPY_C_CONTIGUOUS`` was already deprecated in 1.7.0 though. Not very handy.

**Comment by rgommers** [comment]
Don't merge this yet. I didn't mean to include the commit that changes ``stats/setup.py``. Cython generated code still uses the old API, so that commit probably doesn't work.

Please do comment on the idea of these changes though. Is this OK as a general pattern in ``setup.py``'s?

**Comment by rgommers** [comment]
OK this is good to go now.

File: scipy/_build_utils/__init__.py [hunk 1]
```diff
@@ -1 +1,15 @@
+import numpy as np
 from ._fortran import *
+from scipy.lib._version import NumpyVersion
+
+# Don't use deprecation Numpy C API.  Define this to a fixed version instead of
```
Comment (larsmans) [inline scipy/_build_utils/__init__.py]: deprecation -> deprecated

Comment (rgommers) [inline scipy/_build_utils/__init__.py]: thanks, fixed.

**Comment by argriffing** [comment]
I don't know enough about the Python-C-Fortran interface to meaningfully review this, but TravisCI is giving it the green checkmark!

**Comment by rgommers** [comment]
I'm going to take the lack of comments on the build system changes as an OK for doing all modules this way (please comment sometime soon-ish if you disagree). 

That also implies that it's not necessary to add a command line option to ``f2py`` as done in https://github.com/numpy/numpy/issues/5309.

**Pull request merged**